sound/soc/sof/sof-audio.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/device.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof.h>
+#include "sof-priv.h"
+#include "ops.h"
+
+/*
+ * SOF Driver enumeration.
+ */
+static int sof_machine_check(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+	struct snd_soc_acpi_mach *machine;
+	int ret;
+#endif
+
+	if (plat_data->machine)
+		return 0;
+
+#if !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+	dev_err(sdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
+	return -ENODEV;
+#else
+	/* fallback to nocodec mode */
+	dev_warn(sdev->dev, "No ASoC machine driver found - using nocodec\n");
+	machine = devm_kzalloc(sdev->dev, sizeof(*machine), GFP_KERNEL);
+	if (!machine)
+		return -ENOMEM;
+
+	ret = sof_nocodec_setup(sdev->dev, plat_data, machine,
+				plat_data->desc, plat_data->desc->ops);
+	if (ret < 0)
+		return ret;
+
+	plat_data->machine = machine;
+
+	return 0;
+#endif
+}
+
+static int sof_audio_probe(struct platform_device *pdev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(pdev->dev.parent);
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	struct snd_soc_acpi_mach *machine =
+		(struct snd_soc_acpi_mach *)plat_data->machine;
+	const char *drv_name;
+	int size;
+	int ret;
+
+	/* check machine info */
+	ret = sof_machine_check(sdev);
+	if (ret < 0) {
```

Comment:
@lyakh I must admit I am bit confused by your suggestion to adding another Kconfig here. My drawing illustrates how the machine driver selection is handled. The part that is compiler-check dependent is the no-codec machine drv which happens twice. Once at the beginning if it is forced no-codec mode and once at the end when there's no ACPI match or no platform-specific machine driver available. 

Now which part do you think should be make more explicit?

Reply:
@ranj063 sorry for not being clear. I see this code is also in the current code-base, it isn't new. I mean `sof_machine_check()` will always fail in case
```
#if !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
```
and `plat_data->machine == NULL` right? So, that's a (partially) compile-time decision. So, it might be better done something like
```
#if IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
static int sof_machine_check(struct snd_sof_dev *sdev)
{
	...
}
#else
#define sof_machine_check(x) (-ENODEV)
#endif
```
and then do
```
if (!plat_data->machine && sof_machine_check(sdev) < 0) {
	ret = -ENODEV;
	...
}
```
but that's nitpicking really.